opium/src/middlewares/middleware_cookie_session.ml

```diff
@@ -0,0 +1,117 @@
+let log_src =
+  Logs.Src.create ~doc:"Opium middleware for cookie-based sessions" "opium.session"
+;;
+
+module Logs = (val Logs.src_log log_src : Logs.LOG)
+module Map = Map.Make (String)
+
+module Session = struct
+  type t = string Map.t
+
+  let empty = Map.empty
+
+  let of_yojson yojson =
+    let open Yojson.Safe.Util in
+    let session_list =
+      try Some (yojson |> to_assoc |> List.map (fun (k, v) -> k, to_string v)) with
+      | _ -> None
+    in
+    session_list |> Option.map List.to_seq |> Option.map Map.of_seq
+  ;;
+
+  let to_yojson session =
+    `Assoc (session |> Map.to_seq |> List.of_seq |> List.map (fun (k, v) -> k, `String v))
+  ;;
+
+  let of_json json =
+    try of_yojson (Yojson.Safe.from_string json) with
+    | _ -> None
+  ;;
+
+  let to_json session = session |> to_yojson |> Yojson.Safe.to_string
+
+  let to_sexp session =
+    session
+    |> Map.to_seq
+    |> List.of_seq
+    |> Sexplib0.Sexp_conv.(sexp_of_list (sexp_of_pair sexp_of_string sexp_of_string))
+  ;;
+end
+
+module Env = struct
+  let key : Session.t Context.key = Context.Key.create ("session", Session.to_sexp)
```

Comment:
Since `Session.find` takes a `Request.t` and `Session.set` takes a `Request.t`, what's the advantage of having a middleware that stores the sessions in memory?